pkg/controller/persistentvolume/persistentvolume_recycler_controller.go

```diff
@@ -94,16 +116,61 @@ func NewPersistentVolumeRecycler(kubeClient client.Interface, syncPeriod time.Du
 	return recycler, nil
 }
 
+func (recycler *PersistentVolumeRecycler) shouldRecycle(pv *api.PersistentVolume) error {
+	if pv.Spec.ClaimRef == nil {
+		return fmt.Errorf("Volume does not have a reference to claim")
+	}
+	if pv.Status.Phase == api.VolumeReleased {
+		// This is the first attempt to recycle, allow.
+		return nil
+	}
+	if pv.Status.Phase != api.VolumeFailed {
+		return fmt.Errorf("The volume is not in a final state")
+	}
+
+	// The volume is Failed, should we retry recycling?
+	retryCount := recycler.getRetryCount(pv)
+	glog.V(5).Infof("PersistentVolume[%s] %d attempts failed so far", pv.Name, retryCount)
+	if retryCount >= recycler.maximumRetryCount {
+		glog.V(5).Infof("PersistentVolume[%s] skipping deletion, %d attempts failed", pv.Name, retryCount)
+		return fmt.Errorf("%d attempts failed", retryCount)
+	}
+
+	// retry-count is fine, check the timestamp
+	lastFailureStr, ok := pv.Annotations[pvRecycleLastFailureAnnotationKey]
```

Comment:
what are the benefits/consequences of using annotations to store this versus a local map?

A local map containing counts would remove annotation keys, remove PV updates, etc. but it wouldn't guarantee 3 retries.  A restarted controller manager would necessarily have an empty map and the PV could go 3 more times, even if it attempted recycling twice before the controller manager was restarted.

I don't think the guarantee of X times is important, especially if it makes the code smaller/easier for this feature.